splogger.8
.TH splogger 8
.SH NAME
splogger \- make entries in syslog
.SH SYNOPSIS
.B splogger
[
.I tag
[
.I fac
]
]
.SH DESCRIPTION
.B splogger
reads a series of messages and feeds them to
.BR syslog .
At the front of each message it puts
.I tag
(default:
.BR splogger )
and a numerical timestamp.

.B splogger
checks for
.B alert:
or
.B warning:
at the beginning of each message.
It selects a priority of
LOG_ALERT, LOG_WARNING, or LOG_INFO accordingly.

.B splogger
logs messages with facility
.IR fac .
.I fac
(default: 2)
must be numeric.

.B splogger
converts unprintable characters to question marks.

.B splogger
does not log blank lines.

.B splogger
folds messages after 800 characters,
since
.B syslog
can't handle long messages.
.B splogger
uses a + after the timestamp
to mark folded lines.

Note that the
.B syslog
mechanism is inherently unreliable:
it does not guarantee that messages will be logged.
It is also very slow.
.SH "SEE ALSO"
syslog(3),
logger(8)


CHANGES
20190714 code: removed idedit, install-big. var-qmail packages are no more.
20190708 code: lookup uid/gid at runtime.
20190711 portability problem: explicitly initialize strerr_sys and
         define BIND_8_COMPAT to work around Mac OS X.
20190709 doc: rename INSTALL, SENDMAIL to INSTALL.md, SENDMAIL.md to
         support building on case-insensitive filesystems.
20190715 code: detect and prefer utmpx where available.
20190708 code: use DESTDIR environment variable as root directory in install.
20071130 version: netqmail 1.06
20071130 legal: qmail-1.03 is now in the public domain
20051103 doc: dot-qmail.9 updated for changed (19980613) conf-patrn default, tnx ADM
20040121 version: netqmail 1.05
20040121 code: qmail-smtpd is protected from exceedingly long (eg 2GB)
         header lines
20040121 code: qmail_lspawn, qmail-newmrh, qmail-newu, and qmail-rspawn
         are protected from misbehaving on hosts where the size of an
         integer is not the same as the size of a character pointer
         (eg 64 bit hosts with 32 bit ints)
20031027 version: netqmail 1.04.
20031027 doc: INSTALL points to http://lifewithqmail.org/lwq.html
20031027 doc: qmail.7 identifies installation as netqmail and points to
         http://qmail.org/
20031027 doc: qmail-queue.8 adds explanation of $QMAILQUEUE
20031027 doc: qmail-log.5 adds reference to errors from $QMAILQUEUE script
20031027 doc: FAQ also points to http://cr.yp.to/qmail/faq.html and
         http://qmail.org/
20031027 code: qmail-smtpd identifies itself as netqmail
20031027 code: if $QMAILQUEUE is set, it's invoked instead of qmail-queue
20031024 code: changed errno from int to #include.
20031024 code: fixed .qmail parsing bug.
20031024 code: recognize 0.0.0.0 as a local address.
20031024 code: sendmail's -f flag now overrides environment variables.
19980615 version: qmail 1.03.
19980614 doc: eliminated BIN.setup in favor of a web page.
19980614 code: added other auto* to qmail-showctl output.
19980614 doc: added pointer to immhf.html in qmail-header.5.
19980614 doc: added note to TEST.receive about SMTP command format.
19980614 doc: added FAQ 5.6 on qmail-qmqpd.
19980614 code: removed unused variables in idedit.c.
19980613 code: changed conf-patrn to 002.
19980613 doc: moved SENDMAIL lower in INSTALL.
19980612 code: added install-big.
19980612 code: added BIN.Makefile.
19980612 doc: added BIN.README, BIN.setup.
19980612 code: switched to new install.
19980611 code: added idedit.
19980611 doc: added FAQ 1.3 on $QMAILMFTFILE.
19980611 doc: used bouncesaying in FAQ 5.5.
19980611 code: added except.
19980611 code: added bouncesaying.
19980611 code: allowed unbracketed IP addresses in dns_ipplus() and
         dns_mxip().
19980611 code: allowed spaces after colon in non-bracketed addresses in
         qmail-smtpd.
19980611 doc: cleaned up UPGRADE.
19980528 bug: qmail-smtpd skips first character in rcpthosts() call.
         tnx NND. impact: qmail-smtpd crashes on empty address; and it
	 allows relaying to ""@any.host. fix: use addr.s.
19980515 doc: expanded flock discussion in INSTALL.mbox.
19980515 doc: eliminated flock warning from INSTALL.maildir.
19980515 doc: split REMOVE.binmail out of INSTALL.
19980515 doc: split REMOVE.sendmail out of INSTALL.
19980515 doc: split TEST.deliver and TEST.receive out of INSTALL and
         UPGRADE.
19980515 doc: integrated INSTALL.boot into INSTALL.
19980515 code: cleaned up final output in qmail-qmqpd.c.
19980514 doc: updated procmail notes in INSTALL.mbox. tnx JRM.
19980514 doc: changed FAQ 4.4 to point to INSTALL.mbox for procmail.
         tnx JRM.
19980514 code: separated HELO and EHLO; single-line response for HELO.
         tnx to various people.
19980430 version: qmail 1.02.
19980430 doc: updated SECURITY.
19980430 doc: fixed FAQ 4.9. tnx KB.
19980430 code: changed quote2() to avoid quoting <>.
19980429 code: changed quote_need() to quote empty local parts. tnx HHO.
19980428 doc: added status notes to INSTALL and UPGRADE.
19980428 code: skip setting environment in sendmail.c if PROTO is set.
19980428 code: eliminated recipientmap.
19980428 code: added virtual users to qmail-send.c. tnx RN.
19980428 code: eliminated domain from rewrite() in qmail-send.c.
19980428 code: added binm1, binm1+df, binm2, binm2+df, binm3, binm3+df.
19980428 doc: eliminated most Mailbox references from INSTALL, UPGRADE.
19980428 code: added config-fast.
19980428 code: renamed qmail-config as config.
19980428 code: supported QMAILMFTFILE in qmail-inject.c.
19980428 code: recognized Mail-Followup-To in hfield.c.
19980428 code: replaced rwrecip() with rwappend() in qmail-inject.c.
19980428 code: cleaned up doheaderfile() in qmail-inject.c.
19980426 code: eliminated -type test from qmail-qstat to speed it up.
         tnx FT.
19980421 doc: eliminated remove-rcpthosts comments from FAQ.
19980421 doc: updated FAQ 4.3 to point to Russ Allbery's FAQ. 
19980421 doc: took account of /var/qmail/boot in INSTALL, UPGRADE, and
         INSTALL.vsm.
19980421 code: added /var/qmail/boot, with home, home+df, proc, proc+df.
19980421 doc: skipped make and make man in INSTALL.
19980420 doc: cleaned up mbox description in SENDMAIL.
19980420 code: changed QMQP port to official port 628.
19980402 doc: updated qmsmac references to fastforward.
19980402 doc: replaced qmail-upgrade man page with doc/SENDMAIL.
19980402 code: added qmqpservers output to qmail-showctl.
19980402 code: added qmail-qmqpd.
19980402 code: added qmail-qmqpc.
19980304 code: eliminated del_saywhynoexit in qmail-send.c.
19980304 code: eliminated concurrencynodel in qmail-send.c.
19980222 code: added status() to qmail-send.c.
19980222 code: added concurrencyused to qmail-send.c.
19980128 doc: added note to qmail-getpw.9 about ETXTBSY.
19980127 code: eliminated err_seenmail() in qmail-smtpd.c. tnx PO.
19980126 doc: used $DEFAULT in FAQ where possible.
19980126 code: added DEFAULT in qmail-local.
19980126 code: added -/ to qmail-pw2u.

splogger.c
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>
#include <syslog.h>
#include "error.h"
#include "substdio.h"
#include "subfd.h"
#include "str.h"
#include "scan.h"
#include "fmt.h"

char buf[800]; /* syslog truncates long lines (or crashes); GPACIC */
int bufpos = 0; /* 0 <= bufpos < sizeof(buf) */
int flagcont = 0;
int priority; /* defined if flagcont */
char stamp[FMT_ULONG + FMT_ULONG + 3]; /* defined if flagcont */

void stamp_make()
{
  struct timeval tv;
  char *s;
  gettimeofday(&tv,(struct timezone *) 0);
  s = stamp;
  s += fmt_ulong(s,(unsigned long) tv.tv_sec);
  *s++ = '.';
  s += fmt_uint0(s,(unsigned int) tv.tv_usec,6);
  *s = 0;
}

void flush()
{
  if (bufpos) {
    buf[bufpos] = 0;
    if (flagcont)
      syslog(priority,"%s+%s",stamp,buf); /* logger folds invisibly; GPACIC */
    else {
      stamp_make();
      priority = LOG_INFO;
      if (str_start(buf,"warning:")) priority = LOG_WARNING;
      if (str_start(buf,"alert:")) priority = LOG_ALERT;
      syslog(priority,"%s %s",stamp,buf);
      flagcont = 1;
    }
  }
  bufpos = 0;
}

void main(argc,argv)
int argc;
char **argv;
{
  char ch;

  if (argv[1])
    if (argv[2]) {
      unsigned long facility;
      scan_ulong(argv[2],&facility);
      openlog(argv[1],0,facility << 3);
    }
    else
      openlog(argv[1],0,LOG_MAIL);
  else
    openlog("splogger",0,LOG_MAIL);

  for (;;) {
    if (substdio_get(subfdin,&ch,1) < 1) _exit(0);
    if (ch == '\n') { flush(); flagcont = 0; continue; }
    if (bufpos == sizeof(buf) - 1) flush();
    if ((ch < 32) || (ch > 126)) ch = '?'; /* logger truncates at 0; GPACIC */
    buf[bufpos++] = ch;
  }
}
